Code refactor and improvements

n-ski opened the pull request:
An assortment of improvements that I can't be bothered to split into multiple smaller PRs.

**General changes**:
- Changed the signatures of `private protected` helper methods (`GetCommentSearchResult`, `GetPostSearchResult`, etc). Now instead of there being one method that takes `object` parameter, there is two overloads for each of those methods: one overload takes `JToken` parameter, the other takes `XmlNode` parameter. For classes that override these methods, that makes it clear what kind of object they want to work with. For callers, that makes it clear what kind of object they are allowed to pass.
- Fixed incorrect preview URL of `Search.Post.SearchResult` objects created by `Template.Gelbooru`.
- Moved some of the helper methods from `ABooru` class to separate utility classes in `BooruSharp.Utils` namespace.

**Networking changes**:
- Added `GetResponseAsync(string)` and `GetResponseAsync(Uri)` helper methods in `ABooru` class and pulled `GetResponseAsync(HttpRequestMessage)` helper method from `Pixiv` class up to `ABooru` class.
- Reworked `GetJsonAsync` helper method, now it requires generic type parameter and returns deserialized JSON string into the desired type, intead of the JSON string.
- Fixed `CheckAvailabilityAsync` method, now it properly throws an exception if the service is not available. The method was also made `virtual`, so subclasses can override the default behavior.
- Made sure every `HttpResponseMessage` that is created by calling `GetResponseAsync` method is properly disposed. See [this part](https://archive.is/vhodw#32%) of Steve Gordon's article about `HttpResponseMessage` for more info.
- Moved user agent string to the `TextUtils` class and improved its creation. Now user agent string returned by the utility method looks like this: *BooruSharp/x.y.z*, where *x.y.z* is the current version of BooruSharp package. Utility method also accepts a string parameter that will insert the string in user agent (for example passing `"UnitTests"` to the utility method will return *BooruSharp.UnitTests/x.y.z*).

**Pixiv changes**:
- Updated documentation: added missing exceptions and added remaks about required authorization.
- Refactored a bunch of repeated code into separate methods.
- Changed `CheckUpdateTokenAsync` method visibility to `private` since it is used like a helper method throughout the code, now it also checks that the auth tokens are set.
- Fixed `AddFavoriteAsync` and `RemoveFavoriteAsync` methods not updating auth tokens.
- Added a semaphore to make sure only one login request is being processed in case multiple login/auth requests are made. Semaphore also ensures calls to `AccessToken` and `RefreshToken` properties are blocked until current login request is fully processed and tokens are set/refreshed.

n-ski committed:
Refactor joining and escaping tag arguments

- Added TextUtils class in Utils namespace.

TextUtils class has JoinAndEscape method that combines several strings into one using specified separator (space by default) and escapes it so it can be safely passed as query string in URL.

- Changed ABooru.TagsToString method to use JoinAndEscape method and removed Pixiv.JoinTagsAndEscapeString method to eliminate code duplication.

n-ski committed:
Fix Gelbooru post SearchResult preview URL

n-ski committed:
Refactor code in Pixiv

- Updated documentation: added missing exceptions and added remaks about required authorization
- Refactored a bunch of repeated code into separate methods
- Changed CheckUpdateTokenAsync method visibility to 'private' since it is used like a helper method throughout the code, now it also checks that the auth tokens are set
- Fixed AddFavoriteAsync and RemoveFavoriteAsync methods not updating auth tokens
- Added a semaphore to make sure only one login request is being processed in case multiple login/auth requests are made. Semaphore also ensures calls to AccessToken or RefreshToken properties are blocked until current login request is fully processed and tokens are set.

n-ski committed:
ABooru: Change helper methods signatures

Helper methods with 'private protected' visibility now take either JToken or XmlNode as a parameter. This means caller must ensure the correct type is passed.

n-ski committed:
Improve networking

- Added GetResponseAsync(string) and GetResponseAsync(Uri) helper methods.
- Pulled GetResponseAsync(HttpRequestMessage) helper method from Pixiv class up to ABooru class.
- Reworked GetJsonAsync helper method. Now it requires generic type parameter and will deserialize the JSON string into the desired type immediately.
- Fixed CheckAvailabilityAsync method. Now it properly throws an exception if the service is not available.
- Made sure every HttpResponseMessage is disposed. See this part of Steve Gordon's article about HttpResponseMessage <https://archive.is/vhodw#32%> for more info.

n-ski committed:
Pixiv: Implement IDisposable using the dispose pattern

n-ski committed:
Refactor JSON deserialization

Added a helper method for deserializing HttpContent to desired type

n-ski committed:
Refactor/improve user agent string

- Removed user agent string from ABooru.

- Added util method for creating user agent string, project name and project version are now used to create user agent with optional postfix.

- Changed SetHttpClient unit test: will now test for a regular user agent and a custom one.

n-ski committed:
Refactor string to Rating/TagType enum parsing

- Added RatingUtils and TagUtils. These classes contain methods that parse strings into Rating and TagType enum values.

- Removed ABooru.GetRating and ABooru.StringToTagType methods.

n-ski committed:
Rename XmlEntity class to XmlUtils

Also moved it to Utils namespace.

KoB-Kirito commented:
~~You have to use the second ctor if you want the semaphore to function as a lock (new SemaphoreSlim(1,1)).~~
Me dumb.

KoB-Kirito commented:
Also you don't have to dispose a slim semaphore as it has no unmanaged code unless you access the AvailableWaitHandle, and you don't in this case.
Also having pixiv as a disposable could tempt users to use it in a using block, which is really not how it should be used in most use-cases.

n-ski commented:
> You have to use the second ctor if you want the semaphore to function as a lock (new SemaphoreSlim(1,1)).

That constructor is only ever useful if you want to have some of the "locks" on the semaphore initially taken. `new SemaphoreSlim(1, 1)` and `new SemaphoreSlim(1)` will have the same effect and only one thread will be able to enter semaphore. See [example code](https://dotnetfiddle.net/DJrNGz) here.

> you don't have to dispose a slim semaphore as it has no unmanaged code unless you access the AvailableWaitHandle

Will be addressed, thank you.

n-ski committed:
Pixiv: Unimplement IDisposable

Semaphore doesn't have to be disposed if its AvailableWaitHandle is never accessed.

KoB-Kirito commented:
![firefox_2020-09-16_16-52-41](https://user-images.githubusercontent.com/51691757/93354243-13f8fe00-f83d-11ea-8157-4a4e5e6717bc.png)
?

n-ski commented:
Yeah, the example code is kinda flawed because calls to WriteLine aren't protected. I fixed it.

KoB-Kirito commented on BooruSharp.Others/Pixiv.cs:
```diff
@@ -18,7 +20,7 @@ namespace BooruSharp.Others
     /// <para>https://www.pixiv.net/</para>
     /// </summary>
-    public class Pixiv : ABooru
+    public class Pixiv : ABooru, IDisposable
```
As stated, no IDisposable pls.

KoB-Kirito reviewed (changes requested):
Disposable.

n-ski committed:
Remove unnecessary 'using's

n-ski committed:
Replace '==' operator with 'is' pattern for null checks

See this blog post for the reasoning behind this change
<https://www.gullberg.tk/blog/is-null-versus-null-in-c/>

n-ski committed:
Fixup

n-ski committed:
Refactor booru tag and comment methods

n-ski commented:
Looking back at this, there's too much stuff going on (breaking a lot of public APIs and whatnot). Will probably split those changes into multiple PRs if I can think of a better way to handle them.
